### DISCUSSION TITLE
What is the minimal set of client hints needed for the best analysis result?

### DISCUSSION THREAD

**Discussion by istreeter** [nielsbasjes/yauaa#876] [Q&A] [answered]
Background: we have implemented yauaa into Snowplow Enrich (see snowplow/enrich#639).  And it's working great.

In testing the feature, I get a much better yauaa result (meaning more information) if I configure a site to send the `Sec-CH-UA-Full-Version-List` header.  However, if I configure the site to add any extra header (say `Sec-CH-UA-Arch` or `Sec-CH-UA-Bitness`) then it does not seem to improve the result any further.

Because of this, [in the Snowplow documentation](https://docs.snowplow.io/docs/enriching-your-data/available-enrichments/yauaa-enrichment/) we are currently recommending to just send the `Sec-CH-UA-Full-Version-List` header.  I could not see any reason to document adding the other headers.

So my question is... does this observation match with how you expect the analyzer to work?  Or should I expect that adding extra headers like `Sec-CH-UA-Arch` or `Sec-CH-UA-Bitness` will give a different analysis result?

I know that client hints is an evolving browser feature -- so is it likely that in future we will need to request extra hints to still get the best result?

Thank you for your great work maintaining this very helpful project.

**Comment by nielsbasjes** [comment, marked as answer]
The current situation is that the useragent still has some valuable information which is now duplicated in the client hints.
With browser builders stripping the information from the useragent and moving the info to the client hints you can see the need for the client hint increasing over time in the near future.

So right now you'll see that the client hints in many cases do not have added value because both the UA and CH contain `the same` information.

You mention the Sec-CH-UA-Arch and Sec-CH-UA-Bitness which currently are the only way to detect if a Mac is running an M1 (ARM based) cpu. Without these hints this testcase would actually indicate an Intel cpu instead of ARM: https://github.com/nielsbasjes/yauaa/blob/ecf9ee2fe99b27f61f5cdf8a7a6bd7ed4217f84e/analyzer/src/main/resources/UserAgents/ClientHints-Tests.yaml#L913

So right now (Q1 2023) it is really only some of the Chromium based browsers (Chrome, Edge but not Brave) that do the client hints.

The code that does this augmentation is here https://github.com/nielsbasjes/yauaa/blob/main/analyzer/src/main/java/nl/basjes/parse/useragent/clienthints/ClientHintsAnalyzer.java

And many of the testcases are here:
https://github.com/nielsbasjes/yauaa/blob/main/analyzer/src/main/resources/UserAgents/ClientHints-Tests.yaml

So right now my very simple advise is to request all of these client hints and persist them.

If you feel like finding out what the impact is of removing a client hint you can do perhaps something like download the code and comment the header you want to see the impact of and then run the tests to see what the failure impact is.

**Reply by nielsbasjes** [reply]
A bit deeper dive into the fields and their "added value" and why to keep them or not.

Client hint | Example | Keep it? | Why
-- | -- | -- | -- 
Sec-Ch-Ua | " Not A;Brand";v=“99”, “Chromium”;v=“100”, “Google Chrome”;v=“100” | Yes | You may not have the Sec-Ch-Ua-Full-Version-List with the exact versions.
Sec-Ch-Ua-Full-Version-List | " Not A;Brand";v=“99.0.0.0”, “Chromium”;v=“100.0.4896.75”, “Google Chrome”;v=“100.0.4896.75” | Yes | Is the better variant of Sec-Ch-Ua but it may not be present.
Sec-CH-Ua-Full-Version | “100.0.4896.75” | No | This field is deprecated in the standard. Also this info is also present in the Sec-Ch-Ua-Full-Version-List
Sec-Ch-Ua-Mobile | ?0 | Yes | In the (very rare) case where we cannot determine if it is a phone or tablet this flag determines the end result. 
Sec-Ch-Ua-Platform | “Windows” | Yes | Needed in the very common case of bad version info in the useragent.
Sec-Ch-Ua-Platform-Version | “0.1.0” | Yes | Needed in the very common case of bad version info in the useragent. This "0.1.0" means "Windows 7" because the Platform says "Windows".
Sec-Ch-Ua-Arch | “x86” | Yes | The only way to determine a MacOS system is running an M1/M2 (ARM) instead of an Intel CPU
Sec-Ch-Ua-Bitness | “64” | Yes | Often not present in the useragent
Sec-Ch-Ua-Model | "" | Yes | Often not present in the useragent (brand and device info).
Sec-Ch-Ua-Wow64 | ?0 | No | The only thing this says is that this is Windows (use Platform) and that it is 32 bit software running on a 64 bit system.

To simplify it all I would 
- Ask for all of them (regardless if I say No above) because the browsers are currently in flux in what they put in these fields.
- Persist all headers starting with `Sec-Ch-Ua` along with the `User-Agent`

**Reply by istreeter** [reply]
Thank you this is an extremely helpful answer!

**Reply by dletendre-hc** [reply]
Thank you Niels!

**Reply by nielsbasjes** [reply]
FYI: I've added this to the documentation https://yauaa.basjes.nl/using/clienthints/#which-client-hints-do-i-really-need
